From: Denys Vlasenko <vda.linux@googlemail.com>
Date: Sun, 8 Apr 2018 18:06:24 +0200
Subject: wget: check chunk length for overflowing off_t
Origin: https://git.busybox.net/busybox/commit/?id=8e2174e9bd836e53c8b9c6e00d1bc6e2a718686e
Bug-Debian: https://bugs.debian.org/902724
Bug-Debian-Security: https://security-tracker.debian.org/tracker/CVE-2018-1000517

function                                             old     new   delta
retrieve_file_data                                   428     465     +37
wget_main                                           2386    2389      +3
------------------------------------------------------------------------------
(add/remove: 0/0 grow/shrink: 2/0 up/down: 40/0)               Total: 40 bytes

Signed-off-by: Denys Vlasenko <vda.linux@googlemail.com>
---
 networking/wget.c | 14 +++++++++++---
 1 file changed, 11 insertions(+), 3 deletions(-)

--- a/networking/wget.c
+++ b/networking/wget.c
@@ -663,7 +663,7 @@ static FILE* prepare_ftp_session(FILE **
 	if (ftpcmd("SIZE ", target->path, sfp) == 213) {
 		G.content_len = BB_STRTOOFF(G.wget_buf + 4, NULL, 10);
 		if (G.content_len < 0 || errno) {
-			bb_error_msg_and_die("SIZE value is garbage");
+			bb_error_msg_and_die("bad SIZE value '%s'", G.wget_buf + 4);
 		}
 		G.got_clen = 1;
 	}
@@ -949,11 +949,19 @@ static void NOINLINE retrieve_file_data(
 		if (!G.chunked)
 			break;
 
-		fgets_trim_sanitize(dfp, NULL); /* Eat empty line */
+		/* Each chunk ends with "\r\n" - eat it */
+		fgets_trim_sanitize(dfp, NULL);
  get_clen:
+		/* chunk size format is "HEXNUM[;name[=val]]\r\n" */
 		fgets_trim_sanitize(dfp, NULL);
+		errno = 0;
 		G.content_len = STRTOOFF(G.wget_buf, NULL, 16);
-		/* FIXME: error check? */
+		/*
+		 * Had a bug with inputs like "ffffffff0001f400"
+		 * smashing the heap later. Ensure >= 0.
+		 */
+		if (G.content_len < 0 || errno)
+			bb_error_msg_and_die("bad chunk length '%s'", G.wget_buf);
 		if (G.content_len == 0)
 			break; /* all done! */
 		G.got_clen = 1;
